go/vt/vtgate/evalengine/arithmetic.go

```diff
@@ -215,21 +216,28 @@ func NullsafeCompare(v1, v2 sqltypes.Value, collationID collations.ID) (int, err
 	if v2.IsNull() {
 		return 1, nil
 	}
-	if sqltypes.IsNumber(v1.Type()) || sqltypes.IsNumber(v2.Type()) {
-		lv1, err := newEvalResult(v1)
-		if err != nil {
-			return 0, err
-		}
-		lv2, err := newEvalResult(v2)
-		if err != nil {
-			return 0, err
-		}
-		return compareNumeric(lv1, lv2)
-	}
-	if isByteComparable(v1) && isByteComparable(v2) {
+
+	if isByteComparable(v1.Type()) && isByteComparable(v2.Type()) {
 		return bytes.Compare(v1.ToBytes(), v2.ToBytes()), nil
 	}
-	if v1.IsText() && v2.IsText() && collationID != collations.Unknown {
+
+	typ, err := CoerceTo(v1.Type(), v2.Type()) // TODO systay we should add a method where this decision is done at plantime
+	if err != nil {
+		return 0, err
+	}
+	v1cast, err := castTo(v1, typ)
```

Comment:
based on the TODO, 
what can at least be done is to pass the coerceTo value directly to the `NullsafeCompare` method and calculate that once in the engine.